aws/request/waiter.go

```diff
@@ -30,26 +18,97 @@ type WaitAcceptor struct {
 	Argument string
 }
 
+// ExpectedString returns the string that was expected by the WaitAcceptor
+func (a *WaitAcceptor) ExpectedString() string {
+	switch a.Expected.(type) {
+	case string:
+		return fmt.Sprintf("%q", a.Expected)
+	default:
+		return fmt.Sprintf("%v", a.Expected)
+	}
+}
+
 // A Waiter provides waiting for an operation to complete.
 type Waiter struct {
-	Config
-	Client interface{}
-	Input  interface{}
+	Name        string
+	Delay       WaiterDelay
+	MaxAttempts int
+	Acceptors   []WaitAcceptor
+	Logger      aws.Logger
+
+	RequestOpts []Option
+	NewRequest  func([]Option) (*Request, error)
+}
+
+// A WaiterOption is a function that will update the Waiter value's fields to
+// configure the waiter.
+type WaiterOption func(*Waiter)
+
+// A WaiterDelay is a function that will wait (sleep) between attempts
+// to determine if the target state is reached.
+type WaiterDelay func() error
+
+// WaiterConstDelay returns a waiter delay strategy with a constant delay.
+func WaiterConstDelay(delay time.Duration) WaiterDelay {
+	return func() error {
+		time.Sleep(delay)
+		return nil
+	}
+}
+
+// WaiterWithDelay returns a waiter option to set the delay strategy
+// to be used by the waiter.
+func WaiterWithDelay(fn WaiterDelay) WaiterOption {
```

Comment:
What do you think about something like `WithWaiterDelay`? I think it mixes the best of both suggestions.

Waiter Option examples:
* `WithWaiterDelay`
* `WithWaiterMaxAttempts`
* `WithWaiterLogger`

Request Option examples:
* `WithRetryer`
* `WithLogLevel`